src/adapters/net.cpp

```diff
@@ -324,68 +350,131 @@ namespace net {
-    char essid[IW_ESSID_MAX_SIZE + 1];
+  int wireless_network::scan_cb(struct nl_msg* msg, void* instance) {
+    auto wn = static_cast<wireless_network*>(instance);
+    auto gnlh = static_cast<genlmsghdr*>(nlmsg_data(nlmsg_hdr(msg)));
+    struct nlattr* tb[NL80211_ATTR_MAX + 1];
+    struct nlattr* bss[NL80211_BSS_MAX + 1];
+
+    struct nla_policy bss_policy[NL80211_BSS_MAX + 1]{};
+    bss_policy[NL80211_BSS_TSF].type = NLA_U64;
+    bss_policy[NL80211_BSS_FREQUENCY].type = NLA_U32;
+    bss_policy[NL80211_BSS_BSSID].type = NLA_UNSPEC;
+    bss_policy[NL80211_BSS_BEACON_INTERVAL].type = NLA_U16;
+    bss_policy[NL80211_BSS_CAPABILITY].type = NLA_U16;
+    bss_policy[NL80211_BSS_INFORMATION_ELEMENTS].type = NLA_UNSPEC;
+    bss_policy[NL80211_BSS_SIGNAL_MBM].type = NLA_U32;
+    bss_policy[NL80211_BSS_SIGNAL_UNSPEC].type = NLA_U8;
+    bss_policy[NL80211_BSS_STATUS].type = NLA_U32;
+
+    if (nla_parse(tb, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0), genlmsg_attrlen(gnlh, 0), nullptr) < 0) {
+      return NL_SKIP;
+    }
 
-    struct iwreq req {};
-    req.u.essid.pointer = &essid;
-    req.u.essid.length = sizeof(essid);
-    req.u.essid.flags = 0;
+    if (tb[NL80211_ATTR_BSS] == nullptr) {
+      return NL_SKIP;
+    }
 
-    if (iw_get_ext(socket_fd, m_interface.c_str(), SIOCGIWESSID, &req) != -1) {
-      m_essid = string{essid};
-    } else {
-      m_essid.clear();
+    if (nla_parse_nested(bss, NL80211_BSS_MAX, tb[NL80211_ATTR_BSS], bss_policy) != 0) {
+      return NL_SKIP;
+    }
+
+    if (!wn->associated_or_joined(bss)) {
+      return NL_SKIP;
     }
+
+    wn->parse_essid(bss);
+    wn->parse_frequency(bss);
+    wn->parse_signal(bss);
+    wn->parse_quality(bss);
+
+    return NL_SKIP;
   }
 
   /**
-   * Query for device driver quality values
+   * Check for a connection to a AP
    */
-  void wireless_network::query_quality(const int& socket_fd) {
-    iwrange range{};
-    iwstats stats{};
-
-    // Fill range
-    if (iw_get_range_info(socket_fd, m_interface.c_str(), &range) == -1) {
-      return;
-    }
-    // Fill stats
-    if (iw_get_stats(socket_fd, m_interface.c_str(), &stats, &range, 1) == -1) {
-      return;
+  bool wireless_network::associated_or_joined(struct nlattr** bss) {
+    if (bss[NL80211_BSS_STATUS] == nullptr) {
+      return false;
     }
 
-    // Check if the driver supplies the quality value
-    if (stats.qual.updated & IW_QUAL_QUAL_INVALID) {
-      return;
-    }
-    // Check if the driver supplies the quality level value
-    if (stats.qual.updated & IW_QUAL_LEVEL_INVALID) {
-      return;
-    }
+    auto status = nla_get_u32(bss[NL80211_BSS_STATUS]);
 
-    // Check if the link quality has been uodated
-    if (stats.qual.updated & IW_QUAL_QUAL_UPDATED) {
-      m_linkquality.val = stats.qual.qual;
-      m_linkquality.max = range.max_qual.qual;
-    }
+    return status == NL80211_BSS_STATUS_ASSOCIATED || status == NL80211_BSS_STATUS_IBSS_JOINED;
```

Comment:
`NL80211_BSS_STATUS_AUTHENTICATED` should be tested too.

Reply:
Also a **switch()** or **&** operator will be better to test against multiple cases.